packages/ckeditor5-emoji/src/emojipicker.ts

```diff
@@ -0,0 +1,193 @@
+	declare private _emojiDialog: EmojiDialog;
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get requires() {
+		return [ EmojiLibraryIntegration, ContextualBalloon ] as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'EmojiPicker' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public static override get isOfficialPlugin(): true {
+		return true;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public init(): void {
+		this._balloon = this.editor.plugins.get( ContextualBalloon );
+		this._emojiDialog = this._createEmojiDialog();
+
+		this.editor.ui.componentFactory.add( 'emoji', () => {
+			const button = new ButtonView();
+
+			button.set( {
+				label: this.editor.t( 'Emoji' ),
+				withText: true
+			} );
+
+			button.on( 'execute', () => this.showUI() );
+
+			return button;
+		} );
+	}
+
+	/**
+	 * @internal
+	 */
+	public showUI( initialSearchValue?: string ): void {
+		this._balloon.add( {
+			view: this._emojiDialog,
+			position: this._getBalloonPositionData()
+		} );
+
+		this._emojiDialog.focus();
+
+		if ( initialSearchValue ) {
+			this._emojiDialog.updateSearchValue( initialSearchValue );
+		}
+	}
+
+	/**
+	 * @internal
+	 */
+	private _hideUI() {
+		this._balloon.remove( this._emojiDialog );
+
+		this.editor.editing.view.focus();
+	}
+
+	/**
+	 * @internal
+	 */
+	private _getBalloonPositionData() {
+		const view = this.editor.editing.view;
+		const viewDocument = view.document;
+
+		// Set a target position by converting view selection range to DOM.
+		const target = () => view.domConverter.viewRangeToDom( viewDocument.selection.getFirstRange()! );
+
+		return {
+			target
+		};
+	}
+
+	/**
+	 * @internal
+	 */
+	private _createEmojiDialog() {
+		const editor = this.editor;
+		const emojiDialog = new EmojiDialog( editor );
+
+		emojiDialog.emojiElement.addEventListener( 'emoji-click', event => {
+			const emoji = ( event as EmojiClickEvent ).detail.unicode;
+
+			editor.model.change( writer => {
+				editor.model.insertContent( writer.createText( emoji ) );
+			} );
+
+			this._hideUI();
+		} );
+
+		// Close the dialog while focus is in the editor.
+		editor.keystrokes.set( 'Esc', () => {
+			this._hideUI();
+		} );
+
+		// Close the dialog while focus is in it.
+		emojiDialog.emojiElement.addEventListener( 'keydown', event => {
+			if ( event.key === 'Escape' ) {
+				this._hideUI();
+			}
+		} );
+
+		// Close the dialog when clicking outside of it.
+		clickOutsideHandler( {
+			emitter: emojiDialog,
+			contextElements: [ this._balloon.view.element! ],
+			callback: () => this._hideUI(),
+			activator: () => this._balloon.visibleView === emojiDialog
+		} );
+
+		return emojiDialog;
+	}
+}
+
+class EmojiDialog extends View {
+	public emojiElement: HTMLElement;
+
+	constructor( editor: Editor ) {
+		super( editor.locale );
+
+		this.emojiElement = new Picker( {
+			dataSource: editor.plugins.get( EmojiLibraryIntegration ).localDataUrl
+		} );
+		this.emojiElement.style.height = '400px';
```

Comment:
I'm not sure if hardcoding '400px' is the correct practise in CKEditor 5. We should see how it is done in other plugins.

Reply:
Could we use CSS variables here? This way, an integrator can align the height without hacking the plugin code.